prov/util/src/util_mem_monitor.c

```diff
@@ -86,3 +89,143 @@ void ofi_monitor_unsubscribe(struct ofi_mem_monitor *monitor,
 	       "unsubscribing addr=%p len=%zu\n", addr, len);
 	monitor->unsubscribe(monitor, addr, len);
 }
+
+#if HAVE_UFFD_UNMAP
+
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+
+static int ofi_uffd_register(const void *addr, size_t len, size_t page_size)
+{
+	struct uffdio_register reg;
+	int ret;
+
+	reg.range.start = (uint64_t) (uintptr_t)
+			  ofi_get_page_start(addr, page_size);
+	reg.range.len = ofi_get_page_bytes(addr, len, page_size);
+	reg.mode = UFFDIO_REGISTER_MODE_MISSING;
+	ret = ioctl(uffd.fd, UFFDIO_REGISTER, &reg);
+	if (ret < 0) {
+		if (errno != EINVAL) {
+			FI_WARN(&core_prov, FI_LOG_MR,
+				"ioctl/uffd_unreg: %s\n", strerror(errno));
+		}
+		return -errno;
+	}
+	return 0;
+}
+
+static int ofi_uffd_subscribe(struct ofi_mem_monitor *monitor,
+			      const void *addr, size_t len)
+{
+	int ret;
+
+	assert(monitor == &uffd.monitor);
+	ret = ofi_uffd_register(addr, len, uffd.page_size);
+	if (ret)
+		ret = ofi_uffd_register(addr, len, uffd.hugepage_size);
+	return ret;
+}
+
+static int ofi_uffd_unregister(const void *addr, size_t len, size_t page_size)
+{
+	struct uffdio_range range;
+	int ret;
+
+	range.start = (uint64_t) (uintptr_t)
+		      ofi_get_page_start(addr, page_size);
+	range.len = ofi_get_page_bytes(addr, len, page_size);
+	ret = ioctl(uffd.fd, UFFDIO_UNREGISTER, &range);
+	if (ret < 0) {
+		if (errno != EINVAL) {
+			FI_WARN(&core_prov, FI_LOG_MR,
+				"ioctl/uffd_unreg: %s\n", strerror(errno));
+		}
+		return -errno;
+	}
+	return 0;
+}
+
+static void ofi_uffd_unsubscribe(struct ofi_mem_monitor *monitor,
+				 const void *addr, size_t len)
+{
+	int ret;
+
+	assert(monitor == &uffd.monitor);
+	ret = ofi_uffd_unregister(addr, len, uffd.page_size);
+	if (ret)
+		ret = ofi_uffd_unregister(addr, len, uffd.hugepage_size);
+}
+
+int ofi_uffd_init(void)
+{
+	struct uffdio_api api;
+	int ret;
+
+	ofi_monitor_init(&uffd.monitor);
+	uffd.monitor.subscribe = ofi_uffd_subscribe;
+	uffd.monitor.unsubscribe = ofi_uffd_unsubscribe;
+
+	uffd.page_size = ofi_get_page_size();
+	if (uffd.page_size < 0) {
+		ret = (int) uffd.page_size;
+		goto cleanup;
+	}
+
+	uffd.hugepage_size = ofi_get_hugepage_size();
```

Comment:
gurgle...  This is the case.  See:

https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/mm/hugetlbpage.rst

We'll need to parse through the sizes in /sys/kernel/mm/hugepages/hugepages-XXXXkB.  This will need to be a follow on change.

Reply:
Haha sorry, I had this same problem :)

Reply:
You don't happen to have a solution, do you?  If not, I'll figure out how to iterate over that directory and extract the sizes.

Reply:
No, sorry!